protocol/database-scripting/README.md

```diff
@@ -0,0 +1,102 @@
+Database Scripting
+==============
+
+A protocol for running console scripts on the Healthify app's production database.
+
+Development
+-----------
+
+Develop the scripting story similarly to any other story. In this case, this would
+entail running your script at the rails console of your local server.
+
+While some scripts only lend themselves to being performed via pure SQL, where it
+is possible, you should leverage ActiveRecord to perform our INSERT/UPDATE/DELETE
+operations so that we leverage ActiveRecord validations and callbacks.
+
+While developing the script, it is critical to plan our remediation strategy if
+something does go awry.
+
+One useful approach to consider is to wrap the INSERT/UPDATE/DELETE commands within
+a transaction, and then to assert on certain states of the database after the script
+is complete, rolling back all changes if the assertions fail.
+
+Further, it's important to be able to execute a complete reversal of your changes
+in the event that there are unexpected failures. Ideally, we would write a reversal
+script to immediately reverse our changes if necessary.
+
+For INSERT operations, we would need to retain data about what records ended up
+being INSERT-ed by our script, and then the reversal script would delete those
+records.
+
+For UPDATE/DELETE operations, we would want to dump the data of all the records
+we will be interacting with into a Temp Table (or Tables). Our reversal script would then
+perform a COPY of the data from the Temp Table into the original table (and a deletion
+of updated records in the case of an Update).
+
+Code Review
+-----------
+
+Typically, console scripts are not code that we want to add to the Healthify app's
+code repo. As a result, to solicit Code Review you can put your code in a Github
+[gist](https://help.github.com/articles/about-gists/) and share the gist's link
+with your teammates to get Code Review. Also, be sure to include a link to the gist
+in the description of the corresponding Pivotal Story.
+
+Running database scripts poses a higher risk than deploying a normal feature. Hence,
+it's important to have a higher bar of Code Review than for normal stories. Make sure
+you solicit Code Review from at least two independent reviewers.
```

Comment:
> Do you think it would be valuable to add a paragraph to describe the latter more – ensuring that someone with deep domain experience reviews the SQL?

Yeah I think i discussing those two dimensions to code review would add value.

> Would it make sense to put more emphasis on the a more stringent QA process instead of a more stringent CR process?

Good question. I suppose the reason I was initially inclined to double the CR-ers rather than the QA-ers was because CR seems to have a more defined terminating point, whereas with QA, it's kind of unclear how long you should go about messing around in the app before you're satisfied. That terminating point made me think we might more explicitly get something like double the value from doubling labor on the task, but that could be wrong. I do think you are correct that many bugs will slip by Code Review but may be caught in QA.

@healthify/development Thoughts on adding more CR vs. more QA.

> Also there seems to be an implicit assumption that the reversal script will be less prone to bugs than the forward-change script. That assumption seems shaky to me as well.

We don't need to assume this for there to be value in a reversal script, in that it can give us an extra avenue toward successful restoration, even though sometimes that avenue is suboptimal as well. 

But I also think there is good reason to believe reversal scripts  would be less bug-prone (even if we don't know they're bug free). That reason is that our forward script has to determine a particular set of records to change. The backward script merely needs to retain the info about which set of records were changed then perform a mass COPY or DELETE.